### ISSUE TITLE
Why each epoch spends less time? So does inference?

### ISSUE THREAD

**Issue by Joevaen** [MIC-DKFZ/nnUNet#323]
Hi, fabian.
I noticed that the speed of training one epoch became faster from 550s to 330s. It really makes me surprised and it is an awsome job. Do you mind telling me why the speed raised and how do you  top that? Does inference time decay ?
Best regards,
Joevaen,

**Comment by FabianIsensee** [comment]
Hi,
I am sorry to break this to you but the speed should not improve between epochs. On a properly configures system, the first epoch will be a little bit slower but after that all epochs should have the same speed. If this is not the case, several things can be at fault:
- the CPU was overloaded by some other job and could not keep up with the GPU
- you have an I/O bottleneck but your system managed to cache most of the data after a couple of epochs
- ...
Best,
Fabian

**Comment by Joevaen** [comment]
I am sorry I didn't describe my question clearly. What I represented is not an error encountered but actually a question about knowledge. I mean that I noticed the speed of training after I updated the new nnUNet is faster obviously. Before update, I need to spend 550s to train one epoch, but the time of one epoch decay about ~40% (330s) after I update nnUNet. It does improve a lot!  
I have 2 purposes here:
**_1._** Make sure my nnUNet  process is on a right way which means that time reduction is an advantage of 'new' nnUNet.
**_2._**I would like to know a little points about why the new code of nnUNet can do this if you think it is quite alright, or some direction of speed up training. Of course, it is OK if it is not really proper to expose too much details out. I just want to say that you are really doing a nice job and it is significant to workers and patients. Very thanks to you, all the same. ^-^
Respect,
Joevaen.

**Comment by FabianIsensee** [comment]
Ah sorry, got it!

- When I first implemented nnU-Net I used my own implementation of n-d crossentropy which needed to transpose tensors multiple times to be compatible with the CE provided by pytorch back then. This has now changes and pytorch's native implementation is obviously MUCH faster than my workaround
- I am now computing the class locations for oversampling as part of the preprocessing. This saves a np.argwhere and loading of the entire segmentation maps when sampling training patches. This mostly overcomes I/O bottlenecks but also saves a little bit of CPU.
- Inference now supports mixed precision and is therefore about 2x faster on GPUs with tensorcores (if they support 3d convs).

I am getting epoch times between 133-200s/epoch depending on the dataset and GPU (exception is Hippocampus which is very fast). The epoch times are quite similar between RTX 2080ti, RTX Titan and V100, with the ordering being 2080 ti < V100 < RTX Titan (where < denotes that the card to the right is faster). Best price to performance is 2080ti, followed by RTX Titan. You will achieve this speed with Turing (2080 ti/ Titan RTX) only if you use cudnn 8.0.2 or better. Easiest way to do that would be to use the most recent nvidia pytorch docker. You can also compile pytorch yourself with the correct version of cuDNN (that is what I do)

**Comment by Joevaen** [comment]
Thanks for your clearly answer. I will test that soon! :.)
